core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```
@@ -0,0 +1,413 @@
+        metadata.state = Empty
+        appendMetadataToLog(transactionalId, metadata, responseCallback)
+      }
+    }
+  }
+
+  private def validateTransactionalId(transactionalId: String): Errors =
+    if (transactionalId == null || transactionalId.isEmpty)
+      Errors.INVALID_REQUEST
+    else if (!txnManager.isCoordinatorFor(transactionalId))
+      Errors.NOT_COORDINATOR
+    else if (txnManager.isCoordinatorLoadingInProgress(transactionalId))
+      Errors.COORDINATOR_LOAD_IN_PROGRESS
+    else
+      Errors.NONE
+
+
+  def handleAddPartitionsToTransaction(transactionalId: String,
+                                       pid: Long,
+                                       epoch: Short,
+                                       partitions: collection.Set[TopicPartition],
+                                       responseCallback: TxnMetadataUpdateCallback): Unit = {
+    val errors = validateTransactionalId(transactionalId)
+    if (errors != Errors.NONE)
+      responseCallback(errors)
+    else {
+      // try to update the transaction metadata and append the updated metadata to txn log;
+      // if there is no such metadata treat it as invalid pid mapping error.
+      val (error, newMetadata) = txnManager.getTransactionState(transactionalId) match {
+        case None =>
+          (Errors.INVALID_PID_MAPPING, null)
+
+        case Some(metadata) =>
+          // generate the new transaction metadata with added partitions
+          metadata synchronized {
+            if (metadata.pid != pid) {
+              (Errors.INVALID_PID_MAPPING, null)
+            } else if (metadata.producerEpoch != epoch) {
+              (Errors.INVALID_PRODUCER_EPOCH, null)
+            } else if (metadata.pendingState.isDefined) {
+              // return a retriable exception to let the client backoff and retry
+              (Errors.COORDINATOR_LOAD_IN_PROGRESS, null)
+            } else if (metadata.state != Empty && metadata.state != Ongoing) {
```

Since we return immediately after the `prepareXX` marker is written. In both `initPid` and `addPartition` request handling the metadata state could be in `prepareXX` or `Ongoing`, so we need to handle them in both these two places, better in a consistent way.

In the design doc we said the request will be held on the broker side indefinitely until the current transaction is rolled-forward or rolled back complete. We did this in `initPid`, but here we are directly returning a non-retriable error code. I was planning to do these handling on both places in a separate PR but since @dguy is already adding the logic for `initPid` now maybe we can discuss about that now: after thinking about it a bit more I felt maybe its better to return a retriable exception either immediately (with the new error code we need anyways for the above case) or after some timeout (with the `timeout` error code) on both places, and let the client to back-off and retry. Doing this we can avoid ever increasing the in-memory structures like like the per-broker queues and the purgatory.

Thoughts? @junrao @apurvam @hachikuji @dguy

---

Question more for my own understanding than anything else: 
If `initPid` is holding on to the request on the broker until the transaction is completed, then i think `addPartitions` should never be called when the transactionId is in a `PrepareXX` state. Is that correct?

---

@dguy `initPid` is only called only for the lifetime of a producer, when the producer client completed the current txn and is about to start the next one, it will not call `initPid` but just `addPartitions` as the first request. Only when producer client fails-over the new instance (i.e. in its next life) will send an `initPid` again.

---

The code as it is seems correct to me. If you get `addPartitions`, you should be in an Ongoing state. I also think the `EMPTY` state is invalid in this case.

---

Or is `EMPTY` just signally `NO_ONGOING_TRANSACTION`? If so, it is a valid state.

---

I guess @guozhangwang current point was that a valid state here could also be `PREPARE`, if the previous transaction is still completing, in which case we should return `CONCURRENT_TRANSACTIONS`, and have the client retry with backoff.

---

@guozhangwang @apurvam - cool makes sense now. So i guess we should do the same thing in `initPid` and `addPartitions`, i.e., respond with `CONCURRENT_TRANSACTION` or block until the previous transaction has completed